dataplane/ebpf/src/ingress/tcp.rs

```diff
@@ -116,21 +126,106 @@ pub fn handle_tcp_ingress(ctx: TcContext) -> Result<i32, i64> {
         return Ok(TC_ACT_OK);
     }
 
-    // Calculate l3 cksum
-    // TODO(astoycos) use l3_cksum_replace instead
-    unsafe { (*ip_hdr).check = 0 };
-    let full_cksum = unsafe {
-        bpf_csum_diff(
-            mem::MaybeUninit::zeroed().assume_init(),
+    let mut lb_mapping = LoadBalancerMapping {
+        backend,
+        backend_key,
+        tcp_state,
+    };
+
+    let tcp_hdr_ref = unsafe { tcp_hdr.as_ref().ok_or(TC_ACT_OK)? };
+
+    // If the packet has the RST flag set, it means the connection is being terminated, so remove it
+    // from our map.
+    if tcp_hdr_ref.rst() == 1 {
+        unsafe {
+            LB_CONNECTIONS.remove(&client_key)?;
+        }
+    }
+
+    update_tcp_conns(tcp_hdr_ref, &client_key, &mut lb_mapping)?;
+
+    // inspired by https://github.com/torvalds/linux/blob/master/samples/bpf/tcbpf1_kern.c
+    // update dst_addr in the ip_hdr
+    // recalculate the checksums
+    unsafe {
+        let backend_ip = backend.daddr.to_be();
+
+        let ret = bpf_l4_csum_replace(
+            ctx.skb.skb,
+            TCP_CSUM_OFF,
+            original_daddr as u64,
+            backend_ip as u64,
+            IS_PSEUDO | (mem::size_of_val(&backend_ip) as u64),
+        );
+        if ret != 0 {
+            info!(
+                &ctx, 
+                "Failed to update the TCP checksum after modifying the destination IP");
+            return Ok(TC_ACT_OK);
+        }
+        
+        let ret = bpf_l3_csum_replace(
+            ctx.skb.skb,
+            IP_CSUM_OFF,
+            original_daddr as u64,
+            backend_ip as u64,
+            mem::size_of_val(&backend_ip) as u64,
+        );
+        if ret != 0 {
+            info!(
+                &ctx, 
+                "Failed to update the IP header checksum after modifying the destination IP");
+            return Ok(TC_ACT_OK);
+        }
+
+        let ret = bpf_skb_store_bytes(
+            ctx.skb.skb,
+            IP_DST_OFF,
+            &backend_ip as *const u32 as *const c_void,
+            mem::size_of_val(&backend_ip) as u32,
             0,
-            ip_hdr as *mut u32,
-            Ipv4Hdr::LEN as u32,
+        );
+        if ret != 0 {
+            info!(
+                &ctx,
+                "Failed to update the destination IP address in the packet header");
+            return Ok(TC_ACT_OK);
+        }
+    }
+
+    // update dest in the tcp_hdr
+    // recalculate the checksums
+    unsafe{
+        let backend_port = (backend.dport as u16).to_be();
```

Comment:
Again lets leave as much as we can outside of the `unsafe` block

Reply:
fixed